File: src/converter/plugins/convertParameterDeclaration.ts
import ts, {ParameterDeclaration, SignatureDeclarationBase, TypeNode} from "typescript";
import {createSimplePlugin} from "../plugin.js";
import {CheckCoverageService, checkCoverageServiceKey} from "./CheckCoveragePlugin.js";
import {flatUnionTypes, isNullableType, isNullableUnionType} from "./NullableUnionTypePlugin.js";
import {isNullableStringUnionType} from "./StringUnionTypePlugin.js";
import {ConverterContext} from "../context.js";
import {Render, renderNullable} from "../render.js";
import {escapeIdentifier} from "../../utils/strings.js";
import {AnnotationService, annotationServiceKey} from "./AnnotationPlugin.js";
import {InheritanceModifierService, inheritanceModifierServiceKey} from "./InheritanceModifierPlugin.js";
import {TypeScriptService, typeScriptServiceKey} from "./TypeScriptPlugin.js";

export interface ParameterDeclarationsConfiguration {
    strategy: "function" | "lambda",
    defaultValue?: string
    template: (parameters: string, signature: Signature) => string,
}

export interface ParameterDeclarationConfiguration {
    strategy: "function" | "lambda",
    defaultValue?: string
    inheritanceModifier?: string
    type: TypeNode | undefined,
    nullable: boolean,
}

export interface ParameterInfo {
    parameter: ParameterDeclaration,
    type: TypeNode | undefined,
    nullable: boolean,
    optional: boolean,
}

export type Signature = ParameterInfo[]

export const convertParameterDeclaration = createSimplePlugin((node, context, render) => {
    if (!ts.isParameter(node)) return null

    return convertParameterDeclarationWithFixedType(node, context, render, {
        strategy: ts.isFunctionTypeNode(node.parent)
            ? "lambda"
            : "function",
        type: node.type,
        nullable: false
    })
})

export const convertParameterDeclarations = (
    node: SignatureDeclarationBase,
    context: ConverterContext,
    render: Render,
    configuration: ParameterDeclarationsConfiguration,
) => {
    const {strategy, defaultValue, template} = configuration
    const initialSignature = extractSignature(node)

    const typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    if (strategy === "function") {
        const annotationService = context.lookupService<AnnotationService>(annotationServiceKey)
        const annotations = annotationService?.resolveAnnotations(node, context) ?? []
        const delimiter = `\n\n${annotations.join("\n")}`

        const inheritanceModifierService = context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)
        const signatures = expandUnions(initialSignature, context)

        return signatures
            .map(signature => {
                const inheritanceModifier = inheritanceModifierService?.resolveSignatureInheritanceModifier(node, signature, context) ?? undefined

                const parameters = signature
                    .filter(({parameter}) => !isThisParameter(parameter))
                    .map(({parameter, type, nullable}) => {
                        return convertParameterDeclarationWithFixedType(parameter, context, render, {
                            strategy,
                            defaultValue,
                            inheritanceModifier,
                            type,
                            nullable,
                        })
                    })
                    .filter(Boolean)
                    .join(", ")

                const prefix = signature[0] && isThisParameter(signature[0].parameter)
                    ? `/* ${typeScriptService?.printNode(signature[0].parameter)}, */ `
                    : ""

                return template(prefix + parameters, signature)
            })
            .join(delimiter)
    }

    if (strategy === "lambda") {
        const parameters = node.parameters
            .filter(parameter => !isThisParameter(parameter))
            .map(parameter => {
                return convertParameterDeclarationWithFixedType(parameter, context, render, {
                    strategy,
                    type: parameter.type,
                    nullable: false,
                })
            })
            .filter(Boolean)
            .join(", ")

        const prefix = node.parameters[0] && isThisParameter(node.parameters[0])
            ? `/* ${typeScriptService?.printNode(node.parameters[0])}, */ `
            : ""

        return template(prefix + parameters, initialSignature)
    }

    throw new Error(`Unknown parameter declaration strategy: ${strategy}`)
}

export const convertParameterDeclarationWithFixedType = (
    node: ParameterDeclaration,
    context: ConverterContext,
    render: Render,
    configuration: ParameterDeclarationConfiguration,
) => {
    const {type, strategy, inheritanceModifier} = configuration

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)

    checkCoverageService?.cover(node)
    node.dotDotDotToken && checkCoverageService?.cover(node.dotDotDotToken)
    node.questionToken && checkCoverageService?.cover(node.questionToken)

    let name: string

    if (ts.isIdentifier(node.name)) {
        name = escapeIdentifier(render(node.name))
    } else {
        const parameterIndex = node.parent.parameters.indexOf(node)

        const anonymousParameters = node.parent.parameters
            .filter(parameter => !ts.isIdentifier(parameter.name))

        name = anonymousParameters.length === 1
            ? "options"
            : `param${parameterIndex}`

        checkCoverageService?.deepCover(node.name)
    }

    const isOptional = strategy === "lambda" && Boolean(node.questionToken)

    const defaultValue = configuration.defaultValue ?? "definedExternally"

    const isDefinedExternally = strategy === "function"
        && Boolean(node.questionToken)
        && inheritanceModifier !== "override"
        && defaultValue !== ""

    let renderedType = renderNullable(type, isOptional || configuration.nullable, context, render)

    if (type && node.dotDotDotToken) {
        if (renderedType.startsWith("Array")) {
            renderedType = renderedType.replace(/^Array<(.+)>$/, "$1")
        } else if (renderedType.startsWith("ReadonlyArray")) {
            renderedType = renderedType.replace(/^ReadonlyArray<(.+)>$/, "$1")
        } else if (renderedType.startsWith("js.array.ReadonlyArray")) {
            renderedType = renderedType.replace(/^js.array.ReadonlyArray<(.+)>$/, "$1")
        } else {
            renderedType = `Any? /* ${renderedType} */`
        }
    }


    return `${node.dotDotDotToken ? "vararg " : ""}${name}: ${renderedType}${isOptional ? " /* use undefined for default */" : ""}${isDefinedExternally ? ` = ${defaultValue}` : ""}`
}

const isThisParameter = (parameter: ParameterDeclaration) => (
    ts.isIdentifier(parameter.name)
    && parameter.name.text === "this"
)

const extractSignature = (node: SignatureDeclarationBase) => {
    return node.parameters.map(it => ({
        parameter: it,
        type: it.type,
        nullable: false,
        optional: Boolean(it.questionToken)
    }))
}

const expandUnions = (
    signature: Signature,
    context: ConverterContext,
): Signature[] => {
    const currentSignatures = [signature]

    const checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)

    for (let parameterIndex = 0; parameterIndex < signature.length; parameterIndex++) {
        for (let signatureIndex = 0; signatureIndex < currentSignatures.length; signatureIndex++) {
            const signature = currentSignatures[signatureIndex]
            const {parameter, type, optional} = signature[parameterIndex]

            if (isThisParameter(parameter)) continue

            if (type && isNullableStringUnionType(type, context)) continue

            if (type && ts.isUnionTypeNode(type)) {
                checkCoverageService?.cover(type)

                const generatedSignatures: Signature[] = []

                const nullable = isNullableUnionType(type, context)

                const types = flatUnionTypes(type, context)

                for (const subtype of types) {
                    if (isNullableType(subtype)) {
                        checkCoverageService?.deepCover(type)
                        continue
                    }

                    const generatedSignature: Signature = [...signature]
                    const parameterInfo = {
                        parameter,
                        type: subtype,
                        nullable,
                        optional,
                    }
                    generatedSignature.splice(parameterIndex, 1, parameterInfo)
                    generatedSignatures.push(generatedSignature)
                }

                currentSignatures.splice(signatureIndex, 1, ...generatedSignatures)
            }
        }
    }

    return currentSignatures
}

File: karakum/src/jsMain/kotlin/io/github/sgrishchenko/karakum/extension/plugins/convertParameterDeclaration.kt
package io.github.sgrishchenko.karakum.extension.plugins

import io.github.sgrishchenko.karakum.extension.ConverterContext
import io.github.sgrishchenko.karakum.extension.Render
import io.github.sgrishchenko.karakum.extension.createSimplePlugin
import io.github.sgrishchenko.karakum.extension.plugins.ParameterDeclarationStrategy.Companion.function
import io.github.sgrishchenko.karakum.extension.plugins.ParameterDeclarationStrategy.Companion.lambda
import io.github.sgrishchenko.karakum.extension.renderNullable
import io.github.sgrishchenko.karakum.util.escapeIdentifier
import js.array.ReadonlyArray
import js.objects.JsPlainObject
import seskar.js.JsValue
import typescript.*

sealed external interface ParameterDeclarationStrategy {
    companion object {
        @JsValue("function")
        val function: ParameterDeclarationStrategy

        @JsValue("lambda")
        val lambda: ParameterDeclarationStrategy
    }
}

// TODO: ticket for JsPlainObject
//@OptIn(ExperimentalJsExport::class)
//@JsExport
@JsPlainObject
external interface ParameterDeclarationsConfiguration {
    val strategy: ParameterDeclarationStrategy
    val defaultValue: String?
    val template: (parameters: String, signature: Signature) -> String
}

// TODO: ticket for JsPlainObject
//@OptIn(ExperimentalJsExport::class)
//@JsExport
@JsPlainObject
external interface ParameterDeclarationConfiguration {
    val strategy: ParameterDeclarationStrategy
    val defaultValue: String?
    val inheritanceModifier: String?
    val type: TypeNode?
    val nullable: Boolean
}

// TODO: ticket for JsPlainObject
//@OptIn(ExperimentalJsExport::class)
//@JsExport
@JsPlainObject
external interface ParameterInfo {
    val parameter: ParameterDeclaration
    val type: TypeNode?
    val nullable: Boolean
    val optional: Boolean
}

typealias Signature = ReadonlyArray<ParameterInfo>

val convertParameterDeclaration = createSimplePlugin plugin@{ node: Node, context, render ->
    if (!isParameter(node)) return@plugin null

    convertParameterDeclarationWithFixedType(node, context, render, ParameterDeclarationConfiguration(
        strategy = if (isFunctionTypeNode(node.parent)) lambda else function,
        type = node.type,
        nullable = false,
    ))
}

@OptIn(ExperimentalJsExport::class)
@JsExport
fun convertParameterDeclarations(
    node: SignatureDeclarationBase,
    context: ConverterContext,
    render: Render<Node>,
    configuration: ParameterDeclarationsConfiguration,
): String {
    val strategy = configuration.strategy
    val defaultValue = configuration.defaultValue
    val template = configuration.template
    val initialSignature = extractSignature(node)

    val commentService = context.lookupService<CommentService>(commentServiceKey)
    val typeScriptService = context.lookupService<TypeScriptService>(typeScriptServiceKey)

    if (strategy == function) {
        val annotationService = context.lookupService<AnnotationService>(annotationServiceKey)
        val annotations = annotationService?.resolveAnnotations(node, context) ?: emptyArray()
        val leadingComment = commentService?.renderLeadingComments(node) ?: ""
        val delimiter = "\n\n${leadingComment}${annotations.joinToString(separator = "\n")}"

        val inheritanceModifierService = context.lookupService<InheritanceModifierService>(inheritanceModifierServiceKey)
        val signatures = expandUnions(initialSignature, context)

        return signatures.joinToString(separator = delimiter) { signature ->
            val inheritanceModifier = inheritanceModifierService?.resolveSignatureInheritanceModifier(node, signature, context)

            val parameters = signature
                .filter { !isThisParameter(it.parameter) }
                .map {
                    convertParameterDeclarationWithFixedType(
                        it.parameter, context, render, ParameterDeclarationConfiguration(
                            strategy = strategy,
                            defaultValue = defaultValue,
                            inheritanceModifier = inheritanceModifier,
                            type = it.type,
                            nullable = it.nullable,
                        )
                    )
                }
                .filter { it.isNotEmpty() }
                .joinToString(separator = ", ")

            val firstParameter = signature.getOrNull(0)?.parameter
            val prefix = if (firstParameter != null && isThisParameter(firstParameter)) {
                "/* ${typeScriptService?.printNode(firstParameter)}, */ "
            } else {
                ""
            }

            template(prefix + parameters, signature)
        }
    }

    if (strategy == lambda) {
        val parameters = node.parameters.asArray()
            .filter { parameter -> !isThisParameter(parameter) }
            .map { parameter ->
                convertParameterDeclarationWithFixedType(parameter, context, render, ParameterDeclarationConfiguration(
                    strategy = strategy,
                    type = parameter.type,
                    nullable = false,
                ))
            }
            .filter { it.isNotEmpty() }
            .joinToString(separator = ", ")

        val firstParameter = node.parameters.asArray().getOrNull(0)
        val prefix = if (firstParameter != null && isThisParameter(firstParameter)) {
            "/* ${typeScriptService?.printNode(firstParameter)}, */ "
        } else {
            ""
        }

        return template(prefix + parameters, initialSignature)
    }

    error("Unknown parameter declaration strategy: $strategy")
}

@OptIn(ExperimentalJsExport::class)
@JsExport
fun convertParameterDeclarationWithFixedType(
    node: ParameterDeclaration,
    context: ConverterContext,
    render: Render<Node>,
    configuration: ParameterDeclarationConfiguration,
): String {
    val type = configuration.type
    val strategy = configuration.strategy
    val inheritanceModifier = configuration.inheritanceModifier

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)

    checkCoverageService?.cover(node)
    node.dotDotDotToken?.let { checkCoverageService?.cover(it) }
    node.questionToken?.let { checkCoverageService?.cover(it) }

    val name = if (isIdentifier(node.name)) {
        escapeIdentifier(render(node.name))
    } else {
        checkCoverageService?.deepCover(node.name)

        @Suppress("UNCHECKED_CAST_TO_EXTERNAL_INTERFACE")
        val signature = (node.parent as SignatureDeclarationBase)

        val parameterIndex = signature.parameters.asArray().indexOf(node)

        val anonymousParameters = signature.parameters.asArray()
            .filter { parameter -> !isIdentifier(parameter.name) }

        if (anonymousParameters.size == 1) "options" else "param${parameterIndex}"
    }

    val isOptional = strategy == lambda && node.questionToken != null

    val defaultValue = configuration.defaultValue ?: "definedExternally"

    val isDefinedExternally = strategy == function
        && node.questionToken != null
        && inheritanceModifier !== "override"
        && defaultValue !== ""

    var renderedType = renderNullable(type, isOptional || configuration.nullable, context, render)

    if (type != null && node.dotDotDotToken != null) {
        renderedType = if (renderedType.startsWith("Array")) {
            renderedType.replace("^Array<(.+)>$".toRegex(), "$1")
        } else if (renderedType.startsWith("ReadonlyArray")) {
            renderedType.replace("^ReadonlyArray<(.+)>$".toRegex(), "$1")
        } else if (renderedType.startsWith("js.array.ReadonlyArray")) {
            renderedType.replace("^js.array.ReadonlyArray<(.+)>$".toRegex(), "$1")
        } else {
            "Any? /* $renderedType */"
        }
    }

    return "${if (node.dotDotDotToken != null) "vararg " else ""}${name}: ${renderedType}${if (isOptional) " /* use undefined for default */" else ""}${if (isDefinedExternally) " = $defaultValue" else ""}"
}

private fun isThisParameter(parameter: ParameterDeclaration): Boolean {
    val name = parameter.name

    return isIdentifier(name) && name.text === "this"
}

private fun extractSignature(node: SignatureDeclarationBase): Signature =
    node.parameters.asArray().map {
        ParameterInfo(
            parameter = it,
            type = it.type,
            nullable = false,
            optional = it.questionToken != null
        )
    }.toTypedArray()

private fun expandUnions(
    initialSignature: Signature,
    context: ConverterContext,
): ReadonlyArray<Signature> {
    val currentSignatures = mutableListOf(initialSignature)

    val checkCoverageService = context.lookupService<CheckCoverageService>(checkCoverageServiceKey)

    for (parameterIndex in initialSignature.indices) {
        var signatureIndex = 0
        while (signatureIndex < currentSignatures.size) {
            val signature = currentSignatures[signatureIndex]
            val parameterInfo = signature.getOrNull(parameterIndex)
            if (parameterInfo == null) {
                signatureIndex++
                continue
            }

            val parameter = parameterInfo.parameter
            val type = parameterInfo.type
            val optional = parameterInfo.optional
            val previousOptional = signature.getOrNull(parameterIndex - 1)?.optional ?: false

            if (isThisParameter(parameter)) {
                signatureIndex++
                continue
            }

            if (type != null && isNullableLiteralUnionType(type, context)) {
                signatureIndex++
                continue
            }

            if (type != null && isUnionTypeNode(type)) {
                checkCoverageService?.cover(type)

                val generatedSignatures = mutableListOf<Signature>()

                val nullable = isNullableUnionType(type, context)

                val types = flatUnionTypes(type, context)

                for (subtype in types) {
                    if (isNullableType(subtype)) {
                        checkCoverageService?.deepCover(type)
                        continue
                    }

                    val generatedSignature = signature.toMutableList()
                    val parameterInfo = ParameterInfo(
                        parameter = parameter,
                        type = subtype,
                        nullable = nullable,
                        optional = optional,
                    )
                    generatedSignature.removeAt(parameterIndex)
                    generatedSignature.add(parameterIndex, parameterInfo)
                    generatedSignatures += generatedSignature.toTypedArray()
                }

                if (
                    generatedSignatures.size > 1
                    && optional
                    && !previousOptional
                ) {
                    val strippedSignature = signature.filter { !it.optional }.toTypedArray()
                    generatedSignatures.add(0, strippedSignature)
                }

                currentSignatures.removeAt(signatureIndex)
                currentSignatures.addAll(signatureIndex, generatedSignatures)
            }
            signatureIndex++
        }
    }

    return currentSignatures.toTypedArray()
}
